### PR TITLE
fix(distillers): stop fabricating `Build: ok` for python/ruby script output

### LINKED ISSUE #190: 0.6.5: a `.yaml` path in the command routes any Bash output to a build distiller that fabricates `Build: ok`
## Environment
- **omni 0.6.5** (Homebrew, `fajarhide/tap`) — the installed binary, not source
- macOS (Darwin 25.5)
- Claude Code `PreToolUse`/`PostToolUse` Bash hooks

## Summary

If the **command string** mentions a `.yaml` path, OMNI discards the command's real stdout and substitutes the literal string `Build: ok`. Routing keys off the *file extension in the command*, not off the actual output. The substituted text asserts success, so a failing or negative result is reported as passing.

## Minimal repro — deterministic, isolated to one character

Identical file content, identical command; only the extension differs.

```sh
kubectl kustomize some/overlay > /tmp/av2.yaml    # any sizeable YAML
cp /tmp/av2.yaml /tmp/av2.txt

# A — reads .yaml
python3 -c "
import re
t=open('/tmp/av2.yaml').read()
refs=set(re.findall(r'tokenRef:(?:\n\s+\w+:.*)*?\n\s+name:\s*(\S+)',t))
dests=set()
for m in re.finditer(r'^\s+destination:\n((?:\s{4,}.*\n)+)',t,re.M):
    n=re.search(r'^\s+name:\s*(\S+)',m.group(1),re.M)
    if n: dests.add(n.group(1))
print(' tokenRefs :', sorted(refs))
print(' VSS dests :', sorted(dests))
print(' dangling  :', sorted(refs-dests) or 'NONE')
"

# B — byte-identical input, .txt
#     (same script, s#/tmp/av2.yaml#/tmp/av2.txt#)
```

| Variant | Output the agent receives |
|---|---|
| **A** (`.yaml`) | `Build: ok` |
| **B** (`.txt`) | ` tokenRefs : [...]`<br>` VSS dests : [...]`<br>` dangling  : NONE` |

Reproduced twice in the same session. No OMNI banner accompanied the substitution.

## Ruled out

Narrowed by elimination, so this is not guesswork:

- **Not the output wording.** Printing the same `tokenRefs` / `VSS dests` / `dangling` lines as constants passes through verbatim.
- **Not the output size.** Both variants emit the same ~250 bytes; B passes through.
- **Not the command's Python/regex shape.** The same script over an inline string (no file read) passes through and even correctly reported a *non-empty* dangling set.
- **Not the small-output passthrough threshold.** A ~137-byte variant was correctly passed through with `[OMNI: Passthrough — output too small…]`; this one is above it and gets distilled instead.

The single differing variable is `.yaml` vs `.txt` in the command string.

## Why this one is worse than lost tokens

This was a **verification** command in a GitOps change — it checks whether any `VMUser.tokenRef` fails to resolve to a `VaultStaticSecret` destination. The real answer was `dangling: NONE`. Had it been non-empty, `Build: ok` would have reported a **passing check for a broken manifest**, and the PR would have shipped.

`Build: ok` is not a compression of the input — that string appears nowhere in the real output, and the command is not a build. It is invented, and it is invented in the success direction. That is the failure mode the project positions itself against (#147, "reversible + never-fabricate").

I only caught it because the number looked wrong and I re-ran the check writing to a file and reading it back out-of-band.

## Suspected cause

A distiller is selected by sniffing the command string for a file extension (`.yaml` → build/kustomize distiller) rather than from the tool actually invoked (`python3`). The selected distiller then emits a canned summary regardless of whether the output matches its expected shape.

Same class as the closed #112 (*CloudDistiller re-sniffs the tool from content and misroutes*) and #120 (*Distillation can fabricate a successful result*), but reproducing on 0.6.5 — so the fix does not cover extension-based routing from the command string. Distinct from #159, which is over-compression to a banner rather than substitution with invented text.

## Suggested fixes

1. **Never route on the command string's file extension.** Route on the invoked binary; `python3` has no business reaching a build distiller.
2. **Validate before substituting.** If a distiller's output shares no token with its input, treat it as a misroute and pass through verbatim — cheap and catches the whole fabrication class.
3. **Never emit a bare status verdict** (`ok`, `pass`, `success`) that does not appear in the source text.

### PR DESCRIPTION
**Pull request by fajarhide**
Closes #190

## What

`distill_with_command` listed `python`, `python3` and `ruby` among the build
tools, so any non-test invocation fell through to `BuildDistiller`. That
distiller counts error/warning segments and — finding none — returns the literal
`Build: ok` (9 bytes). A script that printed no error line had its entire stdout
replaced with an invented success verdict, **in the success direction**: the
issue's GitOps verification whose real answer was `dangling: NONE` would have
reported a passing check for a broken manifest. Same class as #170
(`cargo tree` → `Build: ok`), one layer up.

Route `python`/`python3`/`ruby` to `TestDistiller` only when the command names a
test runner (`test`/`pytest`/`unittest`); pass through verbatim otherwise. `pip`,
`rake` and the real compilers stay on the build path.

## Why the reporter's trigger theory is wrong (and why it matters)

The report theorised routing keys on a `.yaml`/`.rb` **file extension**. It does
not — routing is by the first executable (`extract_base_executable`), and both
repros simply started with `python3`/`ruby`. The `.yaml`-vs-`.txt` difference in
the issue body was warm-`~/.omni` scorer nondeterminism, not the extension; with
`OMNI_DB_PATH` isolated, both extensions fabricate identically, and `python3 -c`
with no path fabricates too. An extension-based fix would have missed that.

## Rejected: the token-overlap guard

The issue's fix #2 ("if the distiller's output shares no token with its input,
pass through") also fires on a **legitimate clean build**, where `Build: ok`
correctly shares no token with `Compiling …` / `Finished …` — it would silence
BuildDistiller's one real win to fix a routing bug. The honest fix is to route
on the invoked binary.

## Verified

- Reproduced deterministically at the hook's stage order
  (`score_segments(raw)` → `distill_with_command(raw)`) with an isolated DB:
  before `python3 -c "…"` → `Build: ok`; after → output verbatim. Same for
  `ruby check.rb` and the exact `ruby -c …/bubo.rb` chain from the issue comment.
- Regression test `hands_back_the_output_of_bare_script_interpreters` asserts the
  output survives byte-for-byte and is never `Build: ok`; **confirmed red against
  the pre-fix routing** (throwaway repro returned `"Build: ok"` and panicked),
  green after. `still_routes_real_cargo_builds_and_tests_to_their_distillers`
  guards the other direction.
- Local gates on Homebrew rust **1.94.0** (not CI's pinned **1.97.0**, so clippy
  here does not predict CI): `cargo fmt`, `cargo clippy --all-targets -D warnings`,
  full suite **1,146 passed / 0 failed** — all green.

## Summary

Three distiller routing bugs fixed: (1) `python3`/`ruby` scripts incorrectly produced `Build: ok` instead of their actual output (#190); (2) `git log --oneline` discarded subjects leaving only hashes (#107); (3) `npm`/`pnpm`/etc. output silently dropped from 300 to 30 lines with no marker (#188). Added passthrough arm for bare script interpreters; all three now preserve output verbatim with markers where truncated.

---

## Key Changes

- **`python`/`python3`/`ruby` passthrough**: Moved from `BuildDistiller` routing to verbatim passthrough before build tools check. Scripts that print no error/warning lines previously got fabricated `Build: ok`.
- **`git log --oneline` fix**: Changed `take(7)` to `take_while(|l| !l.is_empty())` — keeps full line including subject, not just hash.
- **JS/TS fallback truncation marker**: Replaced silent `break` with `… and N more` tail; added snapshot for previously unwired `git_log.txt oneline` fixture.

---

## Detailed Breakdown

**src/distillers/mod.rs**
- Added `if matches!(base.as_str(), "python" | "python3" | "ruby")` arm before build tools, returning `input.to_string()` verbatim
- Removed `python`, `python3`, `ruby` from `BuildDistiller` match block (they belong to scripts, not builds)
- Added `hands_back_the_output_of_bare_script_interpreters` regression test covering bare interpreter invocations with edge cases (`-c` args, path segments containing "test")
- Added `still_routes_real_cargo_builds_and_tests_to_their_distillers` guard test ensuring compilers still route correctly
- Collapsed nested truncation loops into single iterator chain; deduplicated `more_tail` helper
- Deleted dead `if filtered_input.len() < input.len()` branch in `JsTsDistiller::distill`

**CHANGELOG.md**
- Documented all three fixes with reproduction details and root-cause analysis
- Clarified that the #190 reporter's extension-based theory was incorrect (routing is by invoked binary, not file extension)

---

## Notes

`cargo clippy` was run on Rust 1.94.0 locally; CI runs 1.97.0 so clippy warnings may differ. Full test suite (1,146 tests) passes locally.

_Last updated: 2026-07-25 02:03:27_

<details open><summary><h3>Greptile Summary</h3></summary>

Fixes a fabrication bug (#190) where `python`, `python3`, and `ruby` executables were routed to `BuildDistiller`, causing arbitrary script output to be silently replaced with the literal `Build: ok` whenever no error/warning lines were found. The PR also addresses a `git log --oneline` subject-stripping bug and a JS/TS fallback that dropped lines without a marker.

- **Script interpreter passthrough**: Adds an early-return block routing all `python`/`python3`/`ruby` invocations to verbatim passthrough, and removes them from the build-tools `matches!` arm, eliminating the `Build: ok` fabrication for arbitrary script output.
- **Test coverage**: The new `hands_back_the_output_of_bare_script_interpreters` test validates both byte-for-byte passthrough and absence of a fabricated verdict, and explicitly covers the `contains(\"test\")` boundary case that an earlier draft mishandled.
- **Deliberate trade-off**: `python -m unittest` now receives verbatim passthrough rather than `TestDistiller` routing; the PR accepts noisier-but-correct output over a distiller that could fabricate a verdict.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

The change is safe to merge — it moves three executables from a distiller that fabricates verdicts to a verbatim passthrough, and the regression test explicitly validates both the byte-for-byte output and the boundary cases that tripped an earlier draft of the fix.

The routing change is narrow and well-bounded: it only affects commands whose base executable is python, python3, or ruby. The pre-fix path sent those to BuildDistiller, which returned a fabricated Build: ok for any script with no error/warning lines — a clear correctness bug. The new path returns the input verbatim, which is always safe. The test covers the critical boundary that an earlier draft mishandled, and the companion test guards the build/test routing on the other side.

**Files Needing Attention:** No files require special attention.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/distillers/mod.rs | Adds verbatim-passthrough branch for python/python3/ruby before the build-tools arm; removes those three executables from the build-tools matches list; adds regression test with boundary cases covering the false-positive contains("test") pattern that was in a prior draft. |
| CHANGELOG.md | Adds a thorough changelog entry documenting the root cause, reproduction steps, rejected alternatives, and deliberate trade-offs for all three bug fixes in this PR. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[distill_with_command] --> B{base executable?}
    B -->|git| C[git::GitDistiller]
    B -->|semgrep/trivy/snyk/...| D[security::SecurityDistiller]
    B -->|gh/hub/glab| E[vcs::VcsDistiller]
    B -->|java/javac/mvn/gradle| F{contains test?}
    F -->|yes| G[test::TestDistiller]
    F -->|no| H[build::BuildDistiller]
    B -->|flutter/dart| I{contains test/analyze?}
    I -->|yes| G
    I -->|no| H
    B -->|python/python3/ruby| J["return input verbatim NEW"]
    B -->|cargo/make/cmake/gcc/pip/...| K{cargo subcommand?}
    K -->|test/bench| G
    K -->|build/check/clippy/...| H
    K -->|tree/metadata/...| J
    K -->|other build tools| L{contains test/pytest?}
    L -->|yes| G
    L -->|no| H
    B -->|vitest/tsc/eslint/npm/...| M[jsts::JsTsDistiller]
    B -->|docker/kubectl/terraform/...| N[cloud::CloudDistiller]
    B -->|ls/find/curl/...| O[system_ops::SystemOpsDistiller]
    B -->|unknown| P[generic::GenericDistiller]

    style J fill:#22c55e,color:#fff
```
</details>

<sub>Reviews (2): Last reviewed commit: ["fix(distillers): drop the test-routing s..."](https://github.com/fajarhide/omni/commit/c0f01a0cb2f85dae7d1302b2680ed2c915bba153) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=47171185)</sub>

**Context used:**

- Context used - CLAUDE.md ([source](https://app.greptile.com/weekndlabs/github/fajarhide/omni/-/custom-context?memory=1653942e-45b5-4251-8409-19e160d03e6e))
- Context used - AGENTS.md ([source](https://app.greptile.com/weekndlabs/github/fajarhide/omni/-/custom-context?memory=5b7b5156-798e-47d0-935c-5cc5cabdcec2))

### CODE REVIEW

File: src/distillers/mod.rs [hunk 1]
```diff
@@ -259,6 +259,23 @@ pub fn distill_with_command(
...
+    // `rake` stay in the build path below: their output is build/task oriented.)
+    if matches!(base.as_str(), "python" | "python3" | "ruby") {
+        if cmd_lower.contains("test")
+            || cmd_lower.contains("pytest")
+            || cmd_lower.contains("unittest")
+        {
+            return test::TestDistiller.distill(segments, input, session);
+        }
+        return input.to_string();
+    }
```
[the remark this answers is not in the document]
Comment (fajarhide) [reply src/distillers/mod.rs:276]: Good catch — but stronger than latent: TestDistiller fabricates `Tests: 1 passed` on that same output, so the substring match just moved #190 to another distiller. Fixed in c0f01a0 by *deleting* the test-routing branch — bare interpreters always pass through, and real `python -m pytest|mypy` are shadowed upstream by their TOML filters. Boundary locked by two added cases (substring-in-`-c`-arg and substring-in-path → passthrough).

File: src/distillers/mod.rs [hunk 2]
```diff
@@ -259,6 +259,23 @@ pub fn distill_with_command(
...
+    // (#190) — the same class as `cargo tree` → `Build: ok` (#170). Only a test
+    // invocation has a distiller here; everything else passes through verbatim,
+    // so we never invent a success verdict for output we cannot parse. (`pip`,
+    // `rake` stay in the build path below: their output is build/task oriented.)
+    if matches!(base.as_str(), "python" | "python3" | "ruby") {
+        if cmd_lower.contains("test")
+            || cmd_lower.contains("pytest")
+            || cmd_lower.contains("unittest")
+        {
+            return test::TestDistiller.distill(segments, input, session);
```
[the remark this answers is not in the document]
Comment (fajarhide) [reply src/distillers/mod.rs]: Agreed — pre-existing debt, not new to this PR. Filed the full `distill_with_command` → `registry.rs` migration as #194 (enhancement/tech-debt) and left the routing here as-is for this fix.
